resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -110,119 +104,401 @@ private[spark] class Client(
+      driverKubernetesSelectors: java.util.Map[String, String],
+      submitServerSecret: Secret): Service = {
+    val driverSubmissionServicePort = new ServicePortBuilder()
+      .withName(SUBMISSION_SERVER_PORT_NAME)
+      .withPort(SUBMISSION_SERVER_PORT)
+      .withNewTargetPort(SUBMISSION_SERVER_PORT)
+      .build()
+    kubernetesClient.services().createNew()
+      .withNewMetadata()
+        .withName(kubernetesAppId)
+        .withLabels(driverKubernetesSelectors)
+        .endMetadata()
+      .withNewSpec()
+        .withType("NodePort")
+        .withSelector(driverKubernetesSelectors)
+        .withPorts(driverSubmissionServicePort)
+        .endSpec()
+      .done()
+  }
+
+  private def createDriverPod(
+      kubernetesClient: KubernetesClient,
+      driverKubernetesSelectors: util.Map[String, String],
+      submitServerSecret: Secret,
+      driverSubmitSslOptions: SSLOptions,
+      sslVolumes: Array[Volume],
+      sslVolumeMounts: Array[VolumeMount],
+      sslEnvs: Array[EnvVar]) = {
+    val containerPorts = buildContainerPorts()
+    val probePingHttpGet = new HTTPGetActionBuilder()
+      .withScheme(if (driverSubmitSslOptions.enabled) "HTTPS" else "HTTP")
+      .withPath("/v1/submissions/ping")
+      .withNewPort(SUBMISSION_SERVER_PORT_NAME)
+      .build()
+    kubernetesClient.pods().createNew()
+      .withNewMetadata()
+        .withName(kubernetesAppId)
+        .withLabels(driverKubernetesSelectors)
+        .endMetadata()
+      .withNewSpec()
+        .withRestartPolicy("OnFailure")
+        .addNewVolume()
+          .withName(SUBMISSION_APP_SECRET_VOLUME_NAME)
+          .withNewSecret()
+            .withSecretName(submitServerSecret.getMetadata.getName)
+            .endSecret()
+          .endVolume()
+        .addToVolumes(sslVolumes: _*)
+        .withServiceAccount(serviceAccount)
+        .addNewContainer()
+          .withName(DRIVER_CONTAINER_NAME)
+          .withImage(driverDockerImage)
+          .withImagePullPolicy("IfNotPresent")
+          .addNewVolumeMount()
+            .withName(SUBMISSION_APP_SECRET_VOLUME_NAME)
+            .withMountPath(secretDirectory)
+            .withReadOnly(true)
+            .endVolumeMount()
+          .addToVolumeMounts(sslVolumeMounts: _*)
+          .addNewEnv()
+            .withName(ENV_SUBMISSION_SECRET_LOCATION)
+            .withValue(s"$secretDirectory/$SUBMISSION_APP_SECRET_NAME")
+            .endEnv()
+          .addNewEnv()
+            .withName(ENV_SUBMISSION_SERVER_PORT)
+            .withValue(SUBMISSION_SERVER_PORT.toString)
+            .endEnv()
+          .addToEnv(sslEnvs: _*)
+          .withPorts(containerPorts.asJava)
+          .withNewReadinessProbe()
+            .withHttpGet(probePingHttpGet)
+            .endReadinessProbe()
+          .withNewLivenessProbe()
+            .withHttpGet(probePingHttpGet)
+            .withInitialDelaySeconds(DRIVER_INITIAL_LIVELINESS_CHECK_DELAY_SECONDS)
```

Comment:
what happens when readinessProbe and livenessProbe are both failing?  does k8s wait for the pod to become ready, or immediately restart?

If this check delay is too short (e.g. a kubelet can't pull the docker image in the 10s window) then I'm worried the driver pod would never get spun up anywhere.

We might need to bump this value to the longest we can imagine a docker image pull taking, or otherwise making it configurable.

I'd probably prefer putting in just the readiness probe and no liveness probe until we get this bit figured out